bus-mapping/src/evm/opcodes/push.rs

```diff
@@ -1,28 +1,28 @@
+use super::Opcode;
 use crate::circuit_input_builder::CircuitInputStateRef;
 use crate::eth_types::GethExecStep;
-// Port this to a macro if possible to avoid defining all the PushN
-use super::Opcode;
 use crate::{
     operation::{StackOp, RW},
     Error,
 };
 
 /// Placeholder structure used to implement [`Opcode`] trait over it corresponding to the
-/// [`OpcodeId::PUSH1`](crate::evm::OpcodeId::PUSH1) `OpcodeId`.
+/// `OpcodeId::PUSH*` `OpcodeId`.
 /// This is responsible of generating all of the associated [`StackOp`]s and place them
 /// inside the trace's [`OperationContainer`](crate::operation::OperationContainer).
 #[derive(Debug, Copy, Clone)]
-pub(crate) struct Push1;
+pub(crate) struct Push<const N: usize>;
 
-impl Opcode for Push1 {
+impl<const N: usize> Opcode for Push<N> {
     fn gen_associated_ops(
         state: &mut CircuitInputStateRef,
         steps: &[GethExecStep],
     ) -> Result<(), Error> {
+        let step = &steps[0];
         state.push_op(StackOp::new(
             RW::WRITE,
             // Get the value and addr from the next step. Being the last position filled with an element in the stack
-            steps[1].stack.last_filled(),
+            step.stack.last_filled().map(|a| a - 1),
```

Comment:
Curious about the consideration not to use `steps[1].stack.last_filled()`.

Reply:
IIUC, that's because Push might run out of gas and therefore there's no way to get the value from the next step. As basically, the next step will contain `STOP` or `REVERT` and will not have added any new values to the stack.

Reply:
Make sense. Thanks for explanation!